src/backend/d3d12/BindGroupD3D12.cpp

```diff
@@ -0,0 +1,153 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "BindGroupD3D12.h"
+#include "BindGroupLayoutD3D12.h"
+#include "BufferD3D12.h"
+#include "SamplerD3D12.h"
+#include "TextureD3D12.h"
+
+#include "D3D12Backend.h"
+
+namespace backend {
+namespace d3d12 {
+
+    BindGroup::BindGroup(Device* device, BindGroupBuilder* builder)
+        : BindGroupBase(builder), device(device) {
+
+        const auto* bgl = ToBackend(GetLayout());
+        const auto& layout = bgl->GetBindingInfo();
+
+        uint32_t cbvOffset = bgl->GetDescriptorOffset(BindGroupLayout::DescriptorType::CBV);
+        uint32_t uavOffset = bgl->GetDescriptorOffset(BindGroupLayout::DescriptorType::UAV);
+        uint32_t srvOffset = bgl->GetDescriptorOffset(BindGroupLayout::DescriptorType::SRV);
+        uint32_t samplerOffset = bgl->GetDescriptorOffset(BindGroupLayout::DescriptorType::Sampler);
+
+        for (size_t binding = 0; binding < layout.mask.size(); ++binding) {
+            if (!layout.mask[binding]) {
+                continue;
+            }
+
+            switch (layout.types[binding]) {
+                case nxt::BindingType::UniformBuffer:
+                    {
+                        auto* view = ToBackend(GetBindingAsBufferView(binding));
+                        descriptors[binding].cbv = view->GetCBVDescriptor();
+                        descriptorOffsets[binding] = cbvOffset++;
+                        cbvUavSrvCount++;
```

Comment:
The information about the CBVUAVSRV count or the Sampler count is already present in the layout. Also the offsets could be computed in the layout instead and returned as a const std::array<uint32_t, kMaxBindingsPerGroup>&. This would avoid some computations and save a bit of memory.